### ISSUE TITLE
SunOS 3.5 doesn't work with the TME SCSI controller

### ISSUE THREAD

**Issue by hbent** [phabrics/nme#8]
I just tried a clean install of SunOS 3.5 on the sun3 "carrera" machine.  I can format and label the disk, but actually booting from the miniroot is a no-go:

```
>b st(0,0,0)
Boot: st(0,0,0)
Boot: sd(0,0,1)vmunix -as
Size: 494840+80288+106616 bytes
Sun UNIX 4.2 Release 3.5EXPORT (GENERIC) #1: Tue Dec 1 19:03:42 PST 1987
Copyright (c) 1986 by Sun Microsystems, Inc.
mem = 16384K (0x1000000)
avail mem = 15761408
Ethernet address = 8:0:20:3:1:50
si0 at vme24d16 200000 vec 0x40
sd0 at si0 slave 0
si0:  si_arb_sel: never won arbitration,
        last phase= 0x86 (Cmd complete MSG)
        csr= 0x1007  bcr= 0  tcr= 0x0
        cbsr= 0x0 (DATA_OUT/FREE)  cdr= 0x0  mr= 0x0  bsr= 0x8
        target= 0, lun= 0    DMA addr= 0x0  count= 0 (0)
        cdb=    0  0  0  0  0  0
sd1 at si0 slave 1
si0:  si_arb_sel: never won arbitration,
        last phase= 0x86 (Cmd complete MSG)
        csr= 0x1007  bcr= 0  tcr= 0x0
        cbsr= 0x0 (DATA_OUT/FREE)  cdr= 0x0  mr= 0x0  bsr= 0x8
        target= 0, lun= 1    DMA addr= 0x0  count= 0 (0)
        cdb=    0  20  0  0  0  0
sd2 at si0 slave 8
si0:  si_arb_sel: never won arbitration,
        last phase= 0x86 (Cmd complete MSG)
        csr= 0x1007  bcr= 0  tcr= 0x0
        cbsr= 0x0 (DATA_OUT/FREE)  cdr= 0x0  mr= 0x0  bsr= 0x8
        target= 1, lun= 0    DMA addr= 0x0  count= 0 (0)
        cdb=    0  0  0  0  0  0
sd3 at si0 slave 9
si0:  si_arb_sel: never won arbitration,
        last phase= 0x86 (Cmd complete MSG)
        csr= 0x1007  bcr= 0  tcr= 0x0
        cbsr= 0x0 (DATA_OUT/FREE)  cdr= 0x0  mr= 0x0  bsr= 0x8
        target= 1, lun= 1    DMA addr= 0x0  count= 0 (0)
        cdb=    0  20  0  0  0  0
st0 at si0 slave 32
st1 at si0 slave 40
zs0 at obio 20000 pri 3
zs1 at obio 0 pri 3
ie0 at obio c0000 pri 3
```

Something is wrong with the SCSI emulation here, but I'm not sure how to go about figuring out what the problem is.  I can provide disk/tape images if they would be helpful.

Here is my configuration:
mainbus0: tme/machine/sun3 my-sun3-idprom.bin
cpu0 at mainbus0: tme/ic/m68020 fpu-type m68881 fpu-compliance unknown fpu-incomplete line-f
obio0 at mainbus0 obio: tme/generic/bus size 16MB
obmem0 at mainbus0 obmem: tme/generic/bus size 4GB
ram0 at obmem0 addr 0x0: tme/host/posix/memory ram 16MB
rom0 at obmem0 addr 0x0fef0000: tme/host/posix/memory rom sun3-carrera-rev-3.0.bin
zs0 at obio0 addr 0x020000 ipl 6: tme/machine/sun3/zs
zs1 at obio0 addr 0x000000 ipl 6: tme/machine/sun3/zs
eeprom0 at obio0 addr 0x040000: tme/host/posix/memory persistent my-sun3-eeprom.bin
clock0 at obio0 addr 0x060000: tme/machine/sun3/clock
mainbus0 memerr at obio0 addr 0x080000 ipl 7
mainbusobie0: tme/machine/sun3/obie
obie0 csr at obio0 addr 0x0c0000 ipl 3
obie0 memory at mainbus0
ie0 at obie0: tme/ic/i82586
bpf1 at ie0: tme/host/bsd/bpf
console0 at zs0 channel A: tme/host/posix/serial device /dev/pts/4 break-carats
si0: tme/machine/sun3/si type vme
si0 csr at vme0 addr 0xff200000 ipl 2
ncr0 at si0: tme/ic/ncr5380
scsibus0 at ncr0: tme/scsi/bus
sd0 at scsibus0: tme/scsi/disk id 0 type acb4000
disk0 at sd0: tme/host/posix/disk file sunos35.img
st0 at scsibus0: tme/scsi/tape id 4 type emulex-mt02 vendor EMULEX product "MT-02 QIC"
tape0 at st0: tme/host/posix/tape
command tape0 load sunos/01 sunos/02 sunos/03 sunos/04 sunos/05 sunos/06 sunos/07 sunos/08 sunos/09
command mainbus0 power up

**Comment by Sothanforax** [comment]
Yeah its a longstanding thing from what I can tell, had the same occur in an installation I attempted some months ago, would be worth looking further into now

**Comment by phabrics** [comment]
Apologies, yes, I meant to look into this. Unfortunately, I had some other stuff on my plate that I wanted to clean up first to make sure they weren't contributing to the problem. Hopefully I will have some more time to look into it soon but any assistance is appreciated, of course. If any of you figure anything out first, let me know. I can assign the bug if you wish.

**Comment by Sothanforax** [comment]
After lots of fiddly debugging, one of the first things wrong/incorrect is that the emulated ACB4000 doesn't have the REZERO UNIT scsi command implemented, this causes a failure on attempting to format the virtual disk, not sure on the miniroot bit yet.

**Comment by Sothanforax** [comment]
Second issue is the showstopper, something to do with lots of bus arbitration failures (or at least that's what the bootloader is complaining about)
alongside an invalid page bus error after the bootloader on the install tape gives up and just dumps you out back into the rom
![Screenshot_2024-11-18_23-10-07](https://github.com/user-attachments/assets/c6101fe0-fe02-477c-8e48-d44053e2a707)

**Comment by Sothanforax** [comment]
oh, oh god I think its all because of unimplemented scsi commands at the disk controller side of things.

**Comment by Sothanforax** [comment]
REZERO UNIT now fakes a successful operation, then MODE SENSE aborts, because it too is also unimplemented. fun
